irc_commands.c

```diff
@@ -57,6 +58,117 @@ static void irc_cmd_pass(irc_t *irc, char **cmd)
 	}
 }
 
+static gboolean irc_sasl_plain_parse(char *input, char **user, char **pass)
+{
+	int len;
+	int i = 0;
+	int part = 0;
+	guint8 *decoded;
+	char *parts[2];
+
+	/* bitlbee's base64_decode wrapper adds an extra null terminator at the end */
+	len = base64_decode(input, &decoded);
+
+	while (i < len && part < 3) {
+		parts[part] = (char *) decoded + i;
+		i += strlen(parts[part]) + 1;
+		part++;
+	}
+
+	/* sanity checks */
+	if (part != 3 || i != (len + 1) || strcmp(parts[0], parts[1]) != 0) {
+		g_free(decoded);
+		return FALSE;
+	} else {
+		*user = g_strdup(parts[0]);
+		*pass = g_strdup(parts[2]);
+		g_free(decoded);
+		return TRUE;
+	}
+}
+
+static gboolean irc_sasl_check_pass(irc_t *irc, char *user, char *pass)
+{
+	storage_status_t status;
+
+	/* just check the password here to be able to reply with useful numerics
+	 * the actual identification will be handled later */
+	status = storage_check_pass(user, pass);
+
+	if (status == STORAGE_OK) {
+		if (!irc->user->nick) {
+			/* set the nick here so we have it for the following numeric */
+			irc->user->nick = g_strdup(user);
+		}
+		irc_send_num(irc, 903, ":Password accepted");
+		return TRUE;
+
+	} else if (status == STORAGE_INVALID_PASSWORD) {
+		irc_send_num(irc, 904, ":Incorrect password");
+	} else if (status == STORAGE_NO_SUCH_USER) {
+		irc_send_num(irc, 904, ":The nick is (probably) not registered");
+	} else {
+		irc_send_num(irc, 904, ":Unknown SASL authentication error");
+	}
+
+	return FALSE;
+}
+
+static void irc_cmd_authenticate(irc_t *irc, char **cmd)
+{
+	/* require the CAP to be enabled, and don't allow authentication before server password */
+	if (!(irc->caps & CAP_SASL) ||
+	    (global.conf->authmode == AUTHMODE_CLOSED && !(irc->status & USTATUS_AUTHORIZED))) {
+		return;
+	}
+
+	if (irc->status & USTATUS_SASL_PLAIN_PENDING) {
+		char *user, *pass;
+
+		irc->status &= ~USTATUS_SASL_PLAIN_PENDING;
+
+		if (!irc_sasl_plain_parse(cmd[1], &user, &pass)) {
+			irc_send_num(irc, 904, ":SASL authentication failed");
+			return;
+		}
+
+		/* let's not support the nick != user case
+		 * if NICK is received after SASL, it will just fail after registration */
+		if (user && irc->user->nick && strcmp(user, irc->user->nick) != 0) {
+			irc_send_num(irc, 902, ":Your SASL username does not match your nickname");
+
+		} else if (irc_sasl_check_pass(irc, user, pass)) {
+			/* and here we do the same thing as the PASS command*/
+			if (irc->status & USTATUS_LOGGED_IN) {
+				char *send_cmd[] = { "identify", pass, NULL };
+				root_command(irc, send_cmd);
+			} else {
+				/* no check_login here - wait for CAP END */
+				irc_setpass(irc, pass);
```

Comment:
Hrm, what's CAP END? I guess I should just read the spec, sorry. This means the SASL auth was done befure the USER and NICK cmd were sent IIRC, so if irc_setpass() set the password now, auto-login magic will happen later?